freqtrade/data/dataprovider.py

```diff
@@ -150,6 +151,80 @@ def _add_external_df(
         self.__producer_pairs_df[producer_name][pair_key] = (dataframe, _last_analyzed)
         logger.debug(f"External DataFrame for {pair_key} from {producer_name} added.")
 
+    def _add_external_candle(
+        self,
```

Comment:
I think that we'll want to have some good, explicit tests for this method.

We knew already that it will contain some slightly more complex logic - and (and that's the main thing) - it's not something where edge-cases are easily testable as it's dependant on upstream / out of process data.